lib/redis_client/ruby_connection/resp3.rb

```diff
@@ -54,6 +54,9 @@ def new_buffer
       String.new(encoding: Encoding::BINARY, capacity: 127)
     end
 
+    SIZE_TO_STRING = Hash.new { |h, k| h[k] = k.to_s }
```

Comment:
I am thinking if this is actually threadsafe? If not, we can just prepopulate an array of size, for example, 100 and use it instead of this hash.

Reply:
Well, thread safe can mean a lot of things, but on MRI it's acceptably safe (may generate some extra strings in case of race but they'll be GC). On Truffle or JRuby however I believe you may get problems.

But either way I don't think this is a good idea.

- That has can grow unbounded and will never be shrunk or reclaimed. That is basically a memory leak, and that hash will have to be marked regularly, slowing down GC pauses.
- These small strings are entirely embeded, have no reference, an never held onto. So from a GC perspective they take very little time. 

Generally speaking allocations aren't necessarily a problem, they might if they cause GC to trigger more often, and GC is slow for other reasons. But here I really don't think it's worth.

Reply:
We can use something like 
```ruby
SIZE_TO_STRING = (0..100).to_a.map(&:to_s)
def size_to_string(size)
  SIZE_TO_STRING[size] || size.to_s
end
```

Assuming sizes most of the times should not be large numbers.

This saves 20Mb per 100k sidekiq jobs, but yes, your points are still valid about these micro optimizations.

Reply:
I wouldn't be surprised if the `size_to_string` method call, plus the hash lookup wouldn't end up being slower than the embedded string allocation.

Reply:
```ruby
require 'benchmark/ips'

SIZE_TO_STRING = (0..100).to_a.map(&:to_s)
def size_to_string(size)
  SIZE_TO_STRING[size] || size.to_s
end

puts "== hit =="
Benchmark.ips do |x|
  x.report("Integer#to_s") { 42.to_s }
  x.report("size_to_string") { size_to_string(42) }
  x.compare!(order: :baseline)
end

puts "== miss =="
Benchmark.ips do |x|
  x.report("Integer#to_s") { 420.to_s }
  x.report("size_to_string") { size_to_string(420) }
  x.compare!(order: :baseline)
end

puts "minor_gc: #{GC.stat(:minor_gc_count)}"
```

Results:
```
$ RUBY_GC_HEAP_INIT_SLOTS=1000000 ruby -v /tmp/int_to_str.rb 
ruby 3.2.0 (2022-12-25 revision a528908271) [arm64-darwin22]
RUBY_GC_HEAP_INIT_SLOTS=1000000 (default value: 10000)
== hit ==
Warming up --------------------------------------
        Integer#to_s     1.314M i/100ms
      size_to_string     1.431M i/100ms
Calculating -------------------------------------
        Integer#to_s     13.236M (± 2.4%) i/s -     67.021M in   5.066487s
      size_to_string     14.139M (± 2.0%) i/s -     71.528M in   5.061079s

Comparison:
        Integer#to_s: 13236033.7 i/s
      size_to_string: 14139247.5 i/s - 1.07x  (± 0.00) faster

== miss ==
Warming up --------------------------------------
        Integer#to_s     1.311M i/100ms
      size_to_string   896.525k i/100ms
Calculating -------------------------------------
        Integer#to_s     13.077M (± 2.7%) i/s -     65.554M in   5.016665s
      size_to_string      8.820M (± 4.1%) i/s -     44.826M in   5.091855s

Comparison:
        Integer#to_s: 13076807.3 i/s
      size_to_string:  8820024.3 i/s - 1.48x  (± 0.00) slower

minor_gc: 242
```

That's a very small gain on hit, bit a big loss on miss. I really don't think it's worth it.

Reply:
Reverted this change.

Made some simple micro benchmark:
```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "benchmark-ips"
end

arr = (1..100).to_a + (101..120).to_a

SIZE_TO_STRING = (0..100).to_a.map(&:to_s)
def size_to_string(size)
  SIZE_TO_STRING[size] || size.to_s
end

Benchmark.ips do |x|
  x.report("to_s") do
    arr.each(&:to_s)
  end

  x.report("cached") do
    arr.each do |e|
      size_to_string(e)
    end
  end

  x.compare!
end
```

```
Warming up --------------------------------------
                to_s     5.879k i/100ms
              cached     7.666k i/100ms
Calculating -------------------------------------
                to_s     58.378k (± 1.5%) i/s -    293.950k in   5.036388s
              cached     76.132k (± 1.5%) i/s -    383.300k in   5.035706s

Comparison:
              cached:    76132.4 i/s
                to_s:    58378.1 i/s - 1.30x  (± 0.00) slower
```